doc/man/man1/sacct.1
.TH sacct "1" "Slurm Commands" "Slurm 26.11" "Slurm Commands"

.SH "NAME"
sacct \- displays accounting data for all jobs and job steps in the
Slurm job accounting log or Slurm database

.SH "SYNOPSIS"
\fBsacct\fR [\fIOPTIONS\fR...]

.SH "DESCRIPTION"
.PP
Accounting information for jobs invoked with Slurm are either logged
in the job accounting log file or saved to the Slurm database, as
configured with the AccountingStorageType parameter.
.PP
The \f3sacct\fP command displays job accounting data stored in the job
accounting log file or Slurm database in a variety of forms for your
analysis. The \f3sacct\fP command displays information on jobs, job
steps, status, and exitcodes by default. You can tailor the output
with the use of the \f3\-\-format=\fP option to specify the fields to
be shown.
.PP
Job records consist of a primary entry for the job as a whole as well as
entries for job steps. The Job Launch page has a more detailed description
of each type of job step.
<https://slurm.schedmd.com/job_launch.html#job_record>
.PP
For the root user, the \f3sacct\fP command displays job accounting
data for all users, although there are options to filter the output to
report only the jobs from a specified user or group.
.PP
For the non\-root user, the \f3sacct\fP command limits the display of
job accounting data to jobs that were launched with their own user
identifier (UID) by default. Data for other users can be displayed
with the \f3\-\-allusers\fP, \f3\-\-user\fP, or \f3\-\-uid\fP options.
.PP
Elapsed time fields are presented as
[days\-]hours:minutes:seconds[.microseconds]. Only 'CPU' fields will
ever have microseconds.

\fBNOTE\fR: If designated, the slurmdbd.conf option PrivateData may further
restrict the accounting data visible to users which are not
SlurmUser, root, or a user with AdminLevel=Admin. See the
slurmdbd.conf man page for additional details on restricting
access to accounting data.

\fBNOTE\fR: The contents of Slurm's database are maintained in lower case.
This may result in some \f3sacct\fP output differing from that of other Slurm
commands.

\fBNOTE\fR: Much of the data reported by \f3sacct\fP has been generated by
the \f2wait3()\fP and \f2getrusage()\fP system calls. Some systems
gather and report incomplete information for these calls;
\f3sacct\fP reports values of 0 for this missing data. See your systems
\f2getrusage (3)\fP man page for information about which data are
actually available on your system.

.SH "OPTIONS"

.TP "10"
\fB\-A\fR, \fB\-\-accounts\fR=<\fIaccount_list\fR>
Displays jobs when a comma separated list of accounts are given as the
argument.
.IP

.TP
\fB\-\-array\fR
Expand job arrays. Display all array tasks on separate lines instead of
displaying groups of array tasks on single lines.
.IP

.TP
\fB\-L\fR, \fB\-\-allclusters\fR
Display jobs ran on all clusters. By default, only jobs ran on the
cluster from where \fBsacct\fR is called are displayed.
.IP

.TP
\fB\-X\fR, \fB\-\-allocations\fR
Only show statistics relevant to the job allocation itself, not taking steps
into consideration.

\fBNOTE\fR: Without including steps, utilization statistics for job
allocation(s) will be reported as zero.
.IP

.TP
\fB\-a\fR, \fB\-\-allusers\fR
Displays all users' jobs when run by user root or if \fBPrivateData\fP is not
configured to \fBjobs\fP.
Otherwise display the current user's jobs
.IP

.TP
\fB\-x\fR, \fB\-\-associations\fR=<\fIassoc_list\fR>
Displays the statistics only for the jobs running under the
association ids specified by the \fBassoc_list\fR operand, which is a
comma\-separated list of association ids. Space characters are not
allowed in the \fBassoc_list\fR. Default is all associations\&.
.IP

.TP
\fB\-B\fR, \fB\-\-batch\-script\fR
This option will print the batch script of job if the job used one. If the job
didn't have a script 'NONE' is output.
.br
\fBNOTE\fR: AccountingStoreFlags=job_script is required for this.
.br
\fBNOTE\fR: Requesting specific job(s) with '\-j' is required for this.
.IP

.TP
\fB\-b\fR, \fB\-\-brief\fR
Displays a brief listing consisting of JobID, State, and ExitCode.
.IP

.TP
\fB\-M\fR, \fB\-\-clusters\fR=<\fIcluster_list\fR>
Displays the statistics only for the jobs started on the clusters
specified by the \fIcluster_list\fR operand, which is a
comma\-separated list of clusters. Space characters are not allowed
in the \fIcluster_list\fR.
A value of '\fIall\fR' will query to run on all clusters.
The default is current cluster you are executing the \fBsacct\fR command on or
all clusters in the federation when executed on a federated cluster.
This option implicitly sets the \fB\-\-local\fR option.
.IP

.TP
\fB\-c\fR, \fB\-\-completion\fR
Use job completion data instead of job accounting. The \fBJobCompType\fR
parameter in the slurm.conf file must be defined to a non\-none option.
Does not support federated cluster information (local data only).
.IP

.TP
\fB\-C\fR, \fB\-\-constraints\fR=<\fIconstraint_list\fR>
Comma separated list to filter jobs based on what constraints/features the job
requested. Multiple options will be treated as 'and' not 'or', so the job would
need all constraints specified to be returned not one or the other.
.IP

.TP
\fB\-\-delimiter\fR=<\fIcharacters\fR>
ASCII characters used to separate the fields when specifying
the \fB\-p\fR or \fB\-P\fR options. The default delimiter
is a '|'. This option is ignored if \fB\-p\fR or \fB\-P\fR options
are not specified.
.IP

.TP
\fB\-D\fR, \fB\-\-duplicates\fR
Allow multiple job records for the same job ID or SLUID to be shown. By default,
if multiple job records for the same job ID or SLUID match the request, only the
most recent one will be shown. Records left in the \fBRESIZING\fR
state by a resize operation are also shown by default. Duplicate records can
result from requeues, federation, resizes or from jobs reaching the
\fBMaxJobId\fR value and resetting. Such records can be distinguished by the
\fBSubmit\fR time stamp, by the \fBOriginalSLUID\fR or by the \fBSLUID\fR
fields.
.IP

.TP
\fB\-E\fR, \fB\-\-endtime\fR=<\fIend_time\fR>
Select jobs in any state before the specified time. If states are
given with the \-s option return jobs in this state before this period.
See the \fBDEFAULT TIME WINDOW\fR section (below) for details about how the
default values for \-\-starttime and \-\-endtime are determined.

Valid time formats are:
.br
HH:MM[:SS][AM|PM]
.br
MMDD[YY][\-HH:MM[:SS]]
.br
MM.DD[.YY][\-HH:MM[:SS]]
.br
MM/DD[/YY][\-HH:MM[:SS]]
.br
YYYY\-MM\-DD[THH:MM[:SS]]
.br
today, midnight, noon, elevenses (11 AM), fika (3 PM), teatime (4 PM)
.br
now[{+|\-}\fIcount\fR[seconds(default)|minutes|hours|days|weeks]]
.IP

.TP
\fB\-\-env\-vars\fR
This option will print the running environment of a batch job, otherwise 'NONE'
is output.
.br
\fBNOTE\fR: AccountingStoreFlags=job_env is required for this.
.br
\fBNOTE\fR: Requesting specific job(s) with '\-j' is required for this.
.IP

.TP
\fB\-\-expand\-patterns\fR
Expand any filename patterns from in \f3StdOut\fP, \f3StdErr\fP and \f3StdIn\fP.
Fields that map to a range of values will use the first value of the range. For
example "%t" for task id will be replaced by "0".
.IP

.TP
\fB\-\-federation\fR
Show jobs from the federation if a member of one.
.IP

.TP
\fB\-f\fR, \fB\-\-file\fR=<\fIfile\fR>
Causes the \f3sacct\fP command to read job accounting data from the
named \f2file\fP instead of the current Slurm job accounting log
file. Only applicable when running the jobcomp/filetxt plugin. Setting this flag
implicitly enables the \-c flag.
.IP

.TP
\fB\-F\fR, \fB\-\-flags\fR=<\fIflag_list\fR>
Comma separated list to filter jobs based on what various ways the jobs were
handled. Current flags are SchedSubmit, SchedMain, SchedBackfill and
StartReceived. SchedSubmit, SchedMain, SchedBackfill describe the scheduler
that started the job.
.IP

.TP
\fB\-o\fR, \fB\-\-format\fR
Comma separated list of fields. (use "\-\-helpformat" for a list of
available fields).

\fBNOTE\fR: When using the format option for listing various fields you can put
a %NUMBER afterwards to specify how many characters should be printed.

e.g. format=name%30 will print 30 characters of field name right
justified. A %\-30 will print 30 characters left justified.

When set, the SACCT_FORMAT environment variable will override the
default format. For example:

SACCT_FORMAT="jobid,user,account,cluster"
.IP

.TP
\fB\-g\fR, \fB\-\-gid\fR=, \fB\-\-group\fR=<\fIgid_or_group_list\fR>
Displays the statistics only for the jobs started with the GID
or the GROUP specified by the \fIgid_list\fR or the \fIgroup_list\fR operand,
which is a comma\-separated list. Space characters are not allowed.
Default is no restrictions.
.IP

.TP
\fB\-h\fR, \fB\-\-help\fR
Displays a general help message.
.IP

.TP
\fB\-e\fR, \fB\-\-helpformat\fR
Print a list of fields that can be specified with the \fB\-\-format\fR option.
.IP

.RS
.PP
.nf
.ft 3
Fields available:

Account             AdminComment        AllocCPUS           AllocNodes
AllocTRES           AssocID             AveCPU              AveCPUFreq
AveDiskRead         AveDiskWrite        AvePages            AveRSS
AveVMSize           Billing             BillingRaw          BlockID
Cluster             Comment             Constraints         ConsumedEnergy
ConsumedEnergyRaw   Container           CPUTime             CPUTimeRAW
DBIndex             DerivedExitCode     Elapsed             ElapsedRaw
Eligible            End                 Exclusive           ExitCode
Extra               FailedNode          Flags               GID
Group               JobID               JobIDRaw            JobName
Layout              Licenses            MaxDiskRead         MaxDiskReadNode
MaxDiskReadTask     MaxDiskWrite        MaxDiskWriteNode    MaxDiskWriteTask
MaxPages            MaxPagesNode        MaxPagesTask        MaxRSS
MaxRSSNode          MaxRSSTask          MaxVMSize           MaxVMSizeNode
MaxVMSizeTask       McsLabel            MinCPU              MinCPUNode
MinCPUTask          NCPUS               NNodes              NodeList
NTasks              OriginalSLUID       OverSubscribe       Partition
Planned             PlannedCPU          PlannedCPURAW       Priority
QOS                 QOSRAW              QOSREQ              Reason
ReqCPUFreq          ReqCPUFreqGov       ReqCPUFreqMax       ReqCPUFreqMin
ReqCPUS             ReqMem              ReqNodes            ReqReservation
ReqTRES             Reservation         ReservationId       Restarts
SegmentSize         SLUID               Start               State
StdErr              StdIn               StdOut              Submit
SubmitLine          Suspended           SystemComment       SystemCPU
Timelimit           TimelimitRaw        TotalCPU            TRESUsageInAve
TRESUsageInMax      TRESUsageInMaxNode  TRESUsageInMaxTask  TRESUsageInMin
TRESUsageInMinNode  TRESUsageInMinTask  TRESUsageInTot      TRESUsageOutAve
TRESUsageOutMax     TRESUsageOutMaxNode TRESUsageOutMaxTask TRESUsageOutMin
TRESUsageOutMinNode TRESUsageOutMinTask TRESUsageOutTot     UID
User                UserCPU             WCKey               WCKeyID
WorkDir
.ft 1
.fi
.RE

\fBNOTE\fR: When using with Ave[RSS|VM]Size or their values in
TRESUsageIn[Ave|Tot]. They represent the average/total of the highest
watermarks over all ranks in the step. When using sstat they represent the
average/total at the moment the command was run.

\fBNOTE\fR: TRESUsage*Min* values represent the lowest highwater mark in the
step.

\fBNOTE\fR: Availability of metrics rely on the \fBjobacct_gather\fP plugin
used. For example the jobacct_gather/cgroup in combination with cgroup/v2 does
not provide Virtual Memory metrics due to limitations in the kernel cgroups
interfaces and will show a 0 for the related fields.

The section titled "Job Accounting Fields" describes these fields.
.IP

.TP
\fB\-J\fR, \fB\-\-job\-name\fR=<\fIjobname_list\fR>
Display jobs that have any of these names.
The list consists of a comma separated list of job names.
.IP

.TP
\fB\-j\fR, \fB\-\-jobs\fR=<\fIjob\fR[.\fIstep\fR]>
Displays information about the specified \fIjob\fR[.\fIstep\fR] or list of
\fIjob\fR[.\fIstep\fR]s.

The \fIjob\fR[.\fIstep\fR]
parameter is a comma\-separated list of jobs.
A SLUID may also be used in place of a numeric job ID.
Space characters are not permitted in this list.
.br
\fBNOTE\fR: A step id of 'batch' will display the information about the
batch step.
.br
By default sacct shows only jobs with Eligible time, but with this
option the non\-eligible will be also shown.
.br
\fBNOTE\fR: If \-\-state is also specified, as non\-eligible are not PD,
then non\-eligible jobs will not be displayed.
See the \fBDEFAULT TIME WINDOW\fR section (below) for details about how this
option changes the default values for \-\-starttime and \-\-endtime.
.IP

.TP
\f3\-\-json\fP, \f3\-\-json\fP=\fIlist\fR, \f3\-\-json\fP=<\fIdata_parser\fR>
Dump job information as JSON using the default data_parser plugin or explicit
data_parser with parameters. Sorting and formatting arguments will be ignored.
.IP

.TP
\fB\-\-local\fR
Show only jobs local to this cluster. Ignore other clusters in this federation
(if any). Overrides \-\-federation.
.IP

.TP
\fB\-l\fR, \fB\-\-long\fR
Equivalent to specifying:

.na
\-\-format=jobid,jobidraw,jobname,partition,maxvmsize,maxvmsizenode,
maxvmsizetask,avevmsize,maxrss,maxrssnode,maxrsstask,averss,maxpages,
maxpagesnode,maxpagestask,avepages,mincpu,mincpunode,mincputask,avecpu,ntasks,
alloccpus,elapsed,state,exitcode,avecpufreq,reqcpufreqmin,reqcpufreqmax,
reqcpufreqgov,reqmem,consumedenergy,maxdiskread,maxdiskreadnode,maxdiskreadtask,
avediskread,maxdiskwrite,maxdiskwritenode,maxdiskwritetask,avediskwrite,
reqtres,alloctres,tresusageinave,tresusageinmax,
tresusageinmaxn,tresusageinmaxt,tresusageinmin,tresusageinminn,tresusageinmint,
tresusageintot,tresusageoutmax,tresusageoutmaxn,
tresusageoutmaxt,tresusageoutave,tresusageouttot
.ad
.IP

.TP
\fB\-i\fR, \fB\-\-nnodes\fR=<\fImin\fR[-\fImax\fR]>
Return jobs that ran on the specified number of nodes.
.IP

.TP
\fB\-I\fR, \fB\-\-ncpus\fR=<\fImin\fR[-\fImax\fR]>
Return jobs that ran on the specified number of cpus.
.IP

.TP
\fB\-\-noconvert\fR
Don't convert units from their original type (e.g. 2048M won't be converted to
2G).
.IP

.TP
\fB\-N\fR, \fB\-\-nodelist\fR=<\fInode_list\fR>
Display jobs that ran on any of these node(s). \fInode_list\fR can be
a ranged string.

\fBNOTE\fR: This is not reliable when nodes are added or removed to Slurm
while jobs are running. Only jobs that started in the specified time range
(\-S, \-E) will be returned.
.IP

.TP
\fB\-n\fR, \fB\-\-noheader\fR
No heading will be added to the output. The default action is to
display a header.
.IP

.TP
\fB\-p\fR, \fB\-\-parsable\fR
Output will be '|' delimited with a '|' at the end. See also the
\fB\-\-delimiter\fR option.
.IP

.TP
\fB\-P\fR, \fB\-\-parsable2\fR
Output will be '|' delimited without a '|' at the end. See also the
\fB\-\-delimiter\fR option.
.IP

.TP
\fB\-r\fR, \fB\-\-partition\fR
Comma separated list of partitions to select jobs and job steps
from. The default is all partitions.
.IP

.TP
\fB\-q\fR, \fB\-\-qos\fR
Only send data about jobs using these qos. Default is all.
.IP

.TP
\fB\-R\fR, \fB\-\-reason\fR=<\fIreason_list\fR>
Comma separated list to filter jobs based on what reason the job wasn't
scheduled outside resources/priority.
.IP

.TP
\fB\-S\fR, \fB\-\-starttime\fR
Select jobs in any state after the specified time. Default is 00:00:00
of the
current day, unless the '\-s' or '\-j' options are used. If the '\-s' option is
used, then the default is 'now'. If states are given with the '\-s' option then
only jobs in this state at this time will be returned. If the '\-j' option is
used, then the default time is Unix Epoch 0. See the \fBDEFAULT TIME WINDOW\fR
section (below) for details about how default values for \-\-starttime and
\-\-endtime are determined.

Valid time formats are:
.br
HH:MM[:SS][AM|PM]
.br
MMDD[YY][\-HH:MM[:SS]]
.br
MM.DD[.YY][\-HH:MM[:SS]]
.br
MM/DD[/YY][\-HH:MM[:SS]]
.br
YYYY\-MM\-DD[THH:MM[:SS]]
.br
today, midnight, noon, elevenses (11 AM), fika (3 PM), teatime (4 PM)
.br
now[{+|\-}\fIcount\fR[seconds(default)|minutes|hours|days|weeks]]
.IP

.TP
\fB\-s\fR, \fB\-\-state\fR=<\fIstate_list\fR>
Selects jobs based on their state during the time period given.
Unless otherwise specified, the start and end time will be the
current time when the \fB\-\-state\fR option is specified and
only currently running jobs can be displayed.
A start and/or end time must be specified to view information about
jobs not currently running.
See the \fBJOB STATE CODES\fR section below for a list of state designators.
Multiple state names may be specified using comma separators. Either the short
or long form of the state name may be used (e.g. \fBCA\fR or \fBCANCELLED\fR)
and the name is case insensitive (i.e. \fRca\fR and \fBCA\fR both work).

\fBNOTE\fR: Note for a job to be selected in the PENDING state it must have
"EligibleTime" in the requested time interval or different from "Unknown". The
"EligibleTime" is displayed by the "scontrol show job" command. For example
jobs submitted with the "\-\-hold" option will have "EligibleTime=Unknown" as
they are pending indefinitely.

\fBNOTE\fR: When specifying states and no start time is given the default
start time is 'now'. This is only when \-j is not used. If \-j is used the
start time will default to 'Epoch'. In both cases if no end time is given it
will default to 'now'. See the \fBDEFAULT TIME WINDOW\fR section (below) for
details about how this option changes the default values for \-\-starttime
and \-\-endtime.
.IP

.TP
\fB\-K\fR, \fB\-\-timelimit\-max\fR
Ignored by itself, but if timelimit_min is set this will be the
maximum timelimit of the range. Default is no restriction.
.IP

.TP
\fB\-k\fR, \fB\-\-timelimit\-min\fR
Only send data about jobs with this timelimit. If used with
timelimit_max this will be the minimum timelimit of the range.
Default is no restriction.
.IP

.TP
\fB\-T\fR, \fB\-\-truncate\fR
Truncate time. So if a job started before \-\-starttime the start time
would be truncated to \-\-starttime. The same for end time and \-\-endtime.
.IP

.TP
\fB\-u\fB, \fB\-\-uid\fR=, \fB\-\-user\fR=<\fIuid_or_user_list\fR>
Use this comma separated list of UIDs or user names to select jobs to
display. By default, the running user's UID is used.
.IP

.TP
\fB\-\-units\fR=[\fBKMGTP\fR]
Display values in specified unit type. Takes precedence over \fB\-\-noconvert\fR
option.
.IP

.TP
\fB\-\-usage\fR
Display a command usage summary.
.IP

.TP
\fB\-\-use\-local\-uid\fR
When displaying UID, sacct uses the UID stored in Slurm's accounting database
by default. Use this command to make Slurm use a system call to get the UID
from the username. This option may be useful in an environment with multiple
clusters and one database where the UIDs aren't the same on all clusters.
.IP

.TP
\fB\-v\fR, \fB\-\-verbose\fR
Primarily for debugging purposes, report the state of various
variables during processing.
.IP

.TP
\fB\-V\fB, \fB\-\-version\fR
Print version.
.IP

.TP
\fB\-W\fR, \fB\-\-wckeys\fR=<\fIwckey_list\fR>
Displays the statistics only for the jobs started on the wckeys
specified by the \f2wckey_list\fP operand, which is a comma\-separated
list of wckey names. Space characters are not allowed in the
\f2wckey_list\fP. Default is all wckeys\&.
.IP

.TP
\fB\-\-whole\-hetjob\fR[=yes|no]
When querying and filtering heterogeneous jobs with \fB\-\-jobs\fR, Slurm will
default to retrieving information about all the components of the job if the
het_job_id (leader id) is selected. If a non\-leader heterogeneous job component
id is selected then only that component is retrieved by default. This behavior
can be changed by using this option. If set to 'yes' (or no argument), then
information about all the components will be retrieved no matter which component
is selected in the job filter. If set to 'no' then only the selected
heterogeneous job component(s) will be retrieved, even when selecting the
leader.
.IP

.TP
\f3\-\-yaml\fP, \f3\-\-yaml\fP=\fIlist\fR, \f3\-\-yaml\fP=<\fIdata_parser\fR>
Dump job information as YAML using the default data_parser plugin or explicit
data_parser with parameters. Sorting and formatting arguments will be ignored.
.IP


NEWS
The contents of NEWS, from slurm-22.05 to the present, has been moved to
CHANGELOG.md and the CHANGELOG directory.


src/sacct/sacct.c
/*****************************************************************************\
 *  sacct.c - job accounting reports for Slurm's jobacct/log plugin
 *****************************************************************************
 *  Copyright (C) 2008-2009 Lawrence Livermore National Security.
 *  Copyright (C) 2005 Hewlett-Packard Development Company, L.P.
 *
 *  This file is part of Slurm, a resource management program.
 *  For details, see <https://slurm.schedmd.com/>.
 *  Please also read the included file: DISCLAIMER.
 *
 *  Slurm is free software; you can redistribute it and/or modify it under
 *  the terms of the GNU General Public License as published by the Free
 *  Software Foundation; either version 2 of the License, or (at your option)
 *  any later version.
 *
 *  In addition, as a special exception, the copyright holders give permission
 *  to link the code of portions of this program with the OpenSSL library under
 *  certain conditions as described in each individual source file, and
 *  distribute linked combinations including the two. You must obey the GNU
 *  General Public License in all respects for all of the code used other than
 *  OpenSSL. If you modify file(s) with this exception, you may extend this
 *  exception to your version of the file(s), but you are not obligated to do
 *  so. If you do not wish to do so, delete this exception statement from your
 *  version.  If you delete this exception statement from all source files in
 *  the program, then also delete it here.
 *
 *  Slurm is distributed in the hope that it will be useful, but WITHOUT ANY
 *  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 *  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 *  details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with Slurm; if not, write to the Free Software Foundation, Inc.,
 *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA.
\*****************************************************************************/

#include "sacct.h"

/*
 * Globals
 */
sacct_parameters_t params;
print_field_t fields[] = {
	{10, "Account", print_fields_str, PRINT_ACCOUNT},
	{15, "AdminComment", print_fields_str, PRINT_ADMIN_COMMENT},
	{10, "AllocCPUS", print_fields_uint, PRINT_ALLOC_CPUS},
	{10, "AllocNodes", print_fields_str, PRINT_ALLOC_NODES},
	{10, "AllocTRES", print_fields_str, PRINT_TRESA},
	{7,  "AssocID", print_fields_uint, PRINT_ASSOCID},
	{10, "AveCPU", print_fields_str, PRINT_AVECPU},
	{10, "AveCPUFreq", print_fields_str, PRINT_ACT_CPUFREQ},
	{14, "AveDiskRead", print_fields_str, PRINT_AVEDISKREAD},
	{14, "AveDiskWrite", print_fields_str, PRINT_AVEDISKWRITE},
	{10, "AvePages", print_fields_str, PRINT_AVEPAGES},
	{10, "AveRSS", print_fields_str, PRINT_AVERSS},
	{10, "AveVMSize", print_fields_str, PRINT_AVEVSIZE},
	{10, "Billing", print_fields_time_from_secs, PRINT_BILLING},
	{10, "BillingRaw", print_fields_uint64, PRINT_BILLING_RAW},
	{16, "BlockID", print_fields_str, PRINT_BLOCKID},
	{10, "Cluster", print_fields_str, PRINT_CLUSTER},
	{14, "Comment", print_fields_str, PRINT_COMMENT},
	{19, "Constraints", print_fields_str, PRINT_CONSTRAINTS},
	{14, "ConsumedEnergy", print_fields_str, PRINT_CONSUMED_ENERGY},
	{17, "ConsumedEnergyRaw", print_fields_uint64,
	 PRINT_CONSUMED_ENERGY_RAW},
	{19, "Container", print_fields_str, PRINT_CONTAINER},
	{10, "CPUTime", print_fields_time_from_secs, PRINT_CPU_TIME},
	{10, "CPUTimeRAW", print_fields_uint64, PRINT_CPU_TIME_RAW},
	{-21, "DBIndex", print_fields_uint64, PRINT_DB_INX},
	{15, "DerivedExitCode", print_fields_str, PRINT_DERIVED_EC},
	{10, "Elapsed", print_fields_time_from_secs, PRINT_ELAPSED},
	{10, "ElapsedRaw", print_fields_uint32, PRINT_ELAPSED_RAW},
	{19, "Eligible", print_fields_date, PRINT_ELIGIBLE},
	{19, "End", print_fields_date, PRINT_END},
	{ 14, "Exclusive", print_fields_str, PRINT_EXCLUSIVE },
	{8,  "ExitCode", print_fields_str, PRINT_EXITCODE},
	{14, "Extra", print_fields_str, PRINT_EXTRA},
	{8,  "FailedNode", print_fields_str, PRINT_FAILED_NODE},
	{19, "Flags", print_fields_str, PRINT_FLAGS},
	{6,  "GID", print_fields_uint, PRINT_GID},
	{9,  "Group", print_fields_str, PRINT_GROUP},
	{-12, "JobID", print_fields_str, PRINT_JOBID},
	{-12, "JobIDRaw", print_fields_str, PRINT_JOBIDRAW},
	{10, "JobName", print_fields_str, PRINT_JOBNAME},
	{9,  "Layout", print_fields_str, PRINT_LAYOUT},
	{10, "Licenses", print_fields_str, PRINT_LICENSES},
	{12, "MaxDiskRead", print_fields_str, PRINT_MAXDISKREAD},
	{15, "MaxDiskReadNode", print_fields_str, PRINT_MAXDISKREADNODE},
	{15, "MaxDiskReadTask", print_fields_uint, PRINT_MAXDISKREADTASK},
	{12, "MaxDiskWrite", print_fields_str, PRINT_MAXDISKWRITE},
	{16, "MaxDiskWriteNode", print_fields_str, PRINT_MAXDISKWRITENODE},
	{16, "MaxDiskWriteTask", print_fields_uint, PRINT_MAXDISKWRITETASK},
	{8,  "MaxPages", print_fields_str, PRINT_MAXPAGES},
	{12, "MaxPagesNode", print_fields_str, PRINT_MAXPAGESNODE},
	{14, "MaxPagesTask", print_fields_uint, PRINT_MAXPAGESTASK},
	{10, "MaxRSS", print_fields_str, PRINT_MAXRSS},
	{10, "MaxRSSNode", print_fields_str, PRINT_MAXRSSNODE},
	{10, "MaxRSSTask", print_fields_uint, PRINT_MAXRSSTASK},
	{10, "MaxVMSize", print_fields_str, PRINT_MAXVSIZE},
	{14, "MaxVMSizeNode", print_fields_str, PRINT_MAXVSIZENODE},
	{14, "MaxVMSizeTask", print_fields_uint, PRINT_MAXVSIZETASK},
	{12, "McsLabel", print_fields_str, PRINT_MCS_LABEL},
	{10, "MinCPU", print_fields_str, PRINT_MINCPU},
	{10, "MinCPUNode", print_fields_str, PRINT_MINCPUNODE},
	{10, "MinCPUTask", print_fields_uint, PRINT_MINCPUTASK},
	{10, "NCPUS", print_fields_uint, PRINT_ALLOC_CPUS},
	{8,  "NNodes", print_fields_uint, PRINT_NNODES},
	{15, "NodeList", print_fields_str, PRINT_NODELIST},
	{8,  "NTasks", print_fields_uint, PRINT_NTASKS},
	{14, "OriginalSLUID", print_fields_sluid, PRINT_ORIGINAL_SLUID},
	{ 14, "OverSubscribe", print_fields_str, PRINT_OVERSUBSCRIBE },
	{10, "Partition", print_fields_str, PRINT_PARTITION},
	{10, "Planned", print_fields_time_from_secs, PRINT_PLANNED},
	{10, "PlannedCPU", print_fields_time_from_secs, PRINT_PLANNED_CPU},
	{13, "PlannedCPURAW", print_fields_uint, PRINT_PLANNED_CPU_RAW},
	{10, "Priority", print_fields_uint, PRINT_PRIO},
	{10, "QOS", print_fields_str, PRINT_QOS},
	{6,  "QOSRAW", print_fields_uint, PRINT_QOSRAW},
	{10, "QOSREQ", print_fields_str, PRINT_QOSREQ},
	{22, "Reason", print_fields_str, PRINT_REASON},
	{10, "ReqCPUFreq", print_fields_str, PRINT_REQ_CPUFREQ_MAX}, /* vestigial */
	{13, "ReqCPUFreqGov", print_fields_str, PRINT_REQ_CPUFREQ_GOV},
	{13, "ReqCPUFreqMax", print_fields_str, PRINT_REQ_CPUFREQ_MAX},
	{13, "ReqCPUFreqMin", print_fields_str, PRINT_REQ_CPUFREQ_MIN},
	{8,  "ReqCPUS", print_fields_uint, PRINT_REQ_CPUS},
	{10, "ReqMem", print_fields_str, PRINT_REQ_MEM},
	{8,  "ReqNodes", print_fields_str, PRINT_REQ_NODES},
	{10, "ReqTRES", print_fields_str, PRINT_TRESR},
	{20, "Reservation",  print_fields_str, PRINT_RESERVATION},
	{8,  "ReservationId",  print_fields_uint, PRINT_RESERVATION_ID},
	{20, "ReqReservation", print_fields_str, PRINT_RESERVATION_REQ},
	{8,  "Restarts", print_fields_uint16, PRINT_RESTART_CNT},
	{11, "SegmentSize", print_fields_uint16, PRINT_SEGMENT_SIZE},
	{14, "SLUID", print_fields_sluid, PRINT_SLUID},
	{19, "Start", print_fields_date, PRINT_START},
	{10, "State", print_fields_str, PRINT_STATE},
	{20, "StdErr", print_fields_str, PRINT_STDERR},
	{20, "StdIn", print_fields_str, PRINT_STDIN},
	{20, "StdOut", print_fields_str, PRINT_STDOUT},
	{19, "Submit", print_fields_date, PRINT_SUBMIT},
	{20, "SubmitLine", print_fields_str, PRINT_SUBMIT_LINE},
	{10, "Suspended", print_fields_time_from_secs, PRINT_SUSPENDED},
	{15, "SystemComment", print_fields_str, PRINT_SYSTEM_COMMENT},
	{10, "SystemCPU", print_fields_str, PRINT_SYSTEMCPU},
	{10, "Timelimit", print_fields_str, PRINT_TIMELIMIT},
	{10, "TimelimitRaw", print_fields_str, PRINT_TIMELIMIT_RAW},
	{10, "TotalCPU", print_fields_str, PRINT_TOTALCPU},
	{14, "TRESUsageInAve", print_fields_str, PRINT_TRESUIA},
	{14, "TRESUsageInMax", print_fields_str, PRINT_TRESUIM},
	{18, "TRESUsageInMaxNode", print_fields_str, PRINT_TRESUIMN},
	{18, "TRESUsageInMaxTask", print_fields_str, PRINT_TRESUIMT},
	{14, "TRESUsageInMin", print_fields_str, PRINT_TRESUIMI},
	{18, "TRESUsageInMinNode", print_fields_str, PRINT_TRESUIMIN},
	{18, "TRESUsageInMinTask", print_fields_str, PRINT_TRESUIMIT},
	{14, "TRESUsageInTot", print_fields_str, PRINT_TRESUIT},
	{15, "TRESUsageOutAve", print_fields_str, PRINT_TRESUOA},
	{15, "TRESUsageOutMax", print_fields_str, PRINT_TRESUOM},
	{19, "TRESUsageOutMaxNode", print_fields_str, PRINT_TRESUOMN},
	{19, "TRESUsageOutMaxTask", print_fields_str, PRINT_TRESUOMT},
	{15, "TRESUsageOutMin", print_fields_str, PRINT_TRESUOMI},
	{19, "TRESUsageOutMinNode", print_fields_str, PRINT_TRESUOMIN},
	{19, "TRESUsageOutMinTask", print_fields_str, PRINT_TRESUOMIT},
	{15, "TRESUsageOutTot", print_fields_str, PRINT_TRESUOT},
	{6,  "UID", print_fields_uint, PRINT_UID},
	{9,  "User", print_fields_str, PRINT_USER},
	{10, "UserCPU", print_fields_str, PRINT_USERCPU},
	{10, "WCKey", print_fields_str, PRINT_WCKEY},
	{10, "WCKeyID", print_fields_uint, PRINT_WCKEYID},
	{20, "WorkDir", print_fields_str, PRINT_WORK_DIR},
	{0,  NULL, NULL, 0}
};

list_t *jobs = NULL;

int main(int argc, char **argv)
{
	int rc = 0;
	data_parser_t *parser = NULL;

	slurm_init(NULL);
	sacct_init();
	parse_command_line(argc, argv);

	if (params.mimetype) {
		rc = data_parser_cli_load(&parser, acct_db_conn, argc, argv,
					  params.mimetype, params.data_parser);
		if (rc || !parser)
			goto cleanup;
	}

	if (!params.mimetype &&
	    !(params.job_cond->flags & JOBCOND_FLAG_SCRIPT) &&
	    !(params.job_cond->flags & JOBCOND_FLAG_ENV))
		print_fields_header(print_fields_list);
	if (get_data() == SLURM_ERROR) {
		if (params.mimetype)
			data_parser_cli_free_ctxt(&parser);
		exit(1);
	}
	if (params.opt_completion)
		do_list_completion();
	else
		do_list(parser);

cleanup:
	if (params.mimetype)
		data_parser_cli_free_ctxt(&parser);

	sacct_fini();

#if MEMORY_LEAK_DEBUG
	slurm_fini();
	log_fini();
#endif

	/* Collapse load errno to 1 to match sacct's historic exit. */
	return (rc ? 1 : 0);
}
